Gnuradio: Enable Darwin Support

lukeadams opened the pull request:
###### Motivation for this change
This PR enables Darwin build support for Gnuradio. 

To accomplish this, it also:
- fixes qwt6_qt4 package on Darwin
- Upgrades libosmocore [0.9.0->0.9.6], with required changes
- minor changes including enableParallelBuilding to others

###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [x] Linux
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Continuation of #26144

Commit:
UHD: enable Darwin build

Commit:
libosmocore: 0.9.0 -> 0.9.6

add new required talloc, python2 inputs; enableParallelBuilding

Mic92 commented on pkgs/applications/misc/gnuradio-ais/default.nix:
```diff
@@ -26,11 +26,15 @@ stdenv.mkDerivation rec {
     done
   '';
 
+  patches = [ ./fix-volk.patch ];
```
Can you also send this patch upstream?

lukeadams replied:
Upstream hasn't had any activity in 2+yr. [Macports seems to have taken the same route and  included the patch with the build definition](https://github.com/macports/macports-ports/tree/master/science/gr-ais)

Mic92 replied:
They might accept straight forward build-fixes though. In the worst case, you still raise awareness and somebody can profit from the changes.

lukeadams replied:
https://github.com/bistromath/gr-ais/pull/18/commits/8502d0252a2a1a9b8d1a71795eaeb5d820684054

Should I update the file to point to my PR commit?

Mic92 replied:
A link would be nice.

lukeadams replied:
Unless you have an objection, I'm going to change the `rev` to point to my PR commit, since it removes the need to include the patch here

lukeadams replied:
It's up

Mic92 commented on pkgs/applications/misc/libosmocore/default.nix:
```diff
@@ -1,30 +1,33 @@
+  ];
   buildInputs = [
-    autoreconfHook pcsclite pkgconfig
+    autoreconfHook pkgconfig git 
```
at least `autoreconfHook`, `pkgconfig` can be moved to `nativeBuildInputs`. Is git really required here?

lukeadams replied:
Nope, git isn't required. I added it while resolving a few compilation errors and never removed it

Mic92 reviewed (commented):
some notes.

bjornfor reviewed (changes requested):
This commit should be split. It does way more than "enabling darwin support". For instance, I don't like that it removes wrapping of the examples, because that breaks them.

bjornfor commented:
Oh, I thought I reviewed only one commit, but it appears github meant it was for the whole PR.

For the remaining changes, please rebase/squash so that there is one logical change per commit.

Commit:
qwt6_qt4: 6.1.2 -> 6.1.3

Commit:
qwt6_qt4: Fix darwin build by adding AGL input

Commit:
qwt6_qt4: enableParallelBuilding

Commit:
gnuradio: document purpose of package inputs

Commit:
gnuradio: use qwt6_qt4 instead of qwt5

Commit:
gnuradio: reorder build inputs

Commit:
gnuradio: add CoreAudio input for Darwin

Commit:
gnuradio: enable Darwin building

Commit:
gnuradio: wrapProgram: set DYLD_FRAMEWORK_PATH

This prevents CoreFoundation-related crashes

Commit:
gnuradio: postInstall: PYTHONPATH / DYLD_FRAMEWORK_PATH notes

Commit:
gnuradio: setupHook: enable composition with nix-shell

Commit:
qwt6_qt4: preFixup: Repair relative reference in darwin framework output

Commit:
gnuradio-ais: Fix Volk library search issue. Fix incorrect version date.

- Issue fixed in my PR (bistromath/gr-ais@8502d02)
- The commit date of the original src is "2015-12-20" not "2016-08-26" (version hasn't changed)

Commit:
rtl-sdr: enable Darwin building; postInstall: only run on Linux

Commit:
gnuradio-{ais,gsm,nacl,osmosdr,rds}: enable Darwin building

Commit:
gnuradio-{ais,gsm,nacl,osmosdr,rds}: enableParallelBuilding

Commit:
libosmocore: enable Darwin building

Commit:
gnuradio-osmosdr: update homepage

lukeadams commented:
>This commit should be split. It does way more than "enabling darwin support". For instance, I don't like that it removes wrapping of the examples, because that breaks them.

I split the gnuradio commit to be more clear

>For the remaining changes, please rebase/squash so that there is one logical change per commit.

I also reordered and split changes to dependencies of gnuradio to make the history more clear.

Hopefully it helps! I apologize for dumping one large commit; I got too aggressive with squashing on my end

bjornfor commented:
@lukeadams: Thank you! That was easy to review. You even split the commits more than I imagined :-)

Does each commit represent a strict improvement and no regressions? I ask because it seems that e.g. the "qwt6_qt4: fixupPhase: Repair relative reference in darwin framework ..." comes a bit late in the sequence. Perhaps squash into "qwt6_qt4: Fix darwin build [by adding AGL input]" or something? (What constitutes a _logical_ change is always up for discussion. A "Fix Darwin build" that does a bunch of stuff is also valid.)

Sorry if I misread the PR, but my point is simply that each commit moves the code base forward linearly. In general, if one commit says "this fixes the build" but then a later one is also needed to _really_ fix the build, then that's not good. Commits should not be split so much that it breaks "git bisect" either.

lukeadams commented:
Glad you liked it!
Regarding the qwt commits:
Darwin building was already enabled but failing, and adding AGL allowed the build to succeed. The fixupPhase basically enabled proper linking to it as a dependency. I can still squash them, but the latter doesn't affect the success/failure of the build.

For the rest of it, I made sure any changes required for darwin were committed *before* enabling darwin building. For example d3d567e has prerequisites for gr-ais darwin compilation, but darwin is not enabled till a later commit. None of the commits break the build, but some may prevent the build from breaking in preparation for darwin building. For some reason Github is showing the commits in the wrong order, which makes it a little confusing to look at. They show up correctly in git log though. Should I try to get the timestamps in order or is it fine?

Is there any way to "test" if git-bisect is broken by any of my commits? Running git bisect start on a clean copy of my branch shows no errors (although I haven't had the chance to use it so I'm not familiar).

thanks for taking the time to review this, this is my first real attempt to contribute to nixpkg and I really appreciate it

bjornfor commented:
@lukeadams: Great. Just wanted to check (commit order).

As for "git bisect" checking. I mostly just _think_ about the commit sequence. To actually verify it you have to do some building. Or a lot, depending on PR rebuild footprint and number of commits.

I run ```nox-review wip --against @{u} && nix-check-before-push.sh```, where nix-check-before-push.sh is

```
nix-env -f . -qa \* --meta --xml --drv-path --show-trace >/dev/null || { echo FAILED; exit 1; }
nix-build pkgs/top-level/release.nix -A tarball || { echo FAILED; exit 1; }
```

And that you'd have to run on each commit. You can do ```git checkout myfeature~N && that_check_command_above``` and change ```N=0...``` until you're back at the commit you branched out from.

I tried to build this PR on top of master and got this failure in the end (NixOS, x86_64-linux):

```
$ nox-review wip --against @{u}
...
209: Test command: /nix/store/1vcp949ka9qnyp6dfv4s9pgjda57vk4x-bash-4.4-p12/bin/sh "/tmp/nix-build-gnuradio-3.7.11.drv-0/gnuradio-3.7.11/build/gr-wavelet/python/wavelet/qa_classify_test.sh"
209: Test timeout computed to be: 9.99988e+06
Scanning dependencies of target gui
209: ....
209: ----------------------------------------------------------------------
209: Ran 4 tests in 0.003s
209:
209: OK
209/209 Test #209: qa_classify ..........................   Passed    0.71 sec

99% tests passed, 1 tests failed out of 209

Total Test time (real) = 208.71 sec

The following tests FAILED:
        196 - qa_qtgui (Failed)
Errors while running CTest
make: *** [Makefile:152: test] Error 8
builder for ‘/nix/store/4syq5hckgf8vdd9f0np395gqxri5cm6r-gnuradio-3.7.11.drv’ failed with exit code 2
error: build of ‘/nix/store/4syq5hckgf8vdd9f0np395gqxri5cm6r-gnuradio-3.7.11.drv’ failed
The invocation of "nix-build -A gnuradio-gsm -A gnuradio-osmosdr -A gnuradio-nacl -A libosmocore -A inspectrum -A gplates -A qwt6_qt4 -A gnuradio -A gnuradio-rds -A gnuradio-with-packages -A gnuradio-ais -A gqrx /home/bfo/proj/code/forks/nixpkgs-master" failed
```

Some small nitpick I found after a 2nd look. You don't _have_ to fix them, but it'd be nice.
* Commit style for version bumps is "qwt6_qt4: 6.1.2 -> 6.1.3", not "qwt6_qt4: [6.1.2->6.1.3]"
* Some trailing whitespace added various places. Spot them with "git log -p".

lukeadams commented:
Extremely insightful! I'll play around with git bisect today when I get the chance.

Regarding the qa_qtgui failure: it seems related to switching to qwt6_qt4, and is the reason for [this snippet](https://github.com/lukeadams/nixpkgs/blob/3a8816095824d5d4e44487e8df62c03cd3299208/pkgs/applications/misc/gnuradio/default.nix#L67-L68) (which fixes the issue on Darwin):
```
cmakeFlags =                             
    stdenv.lib.optionals stdenv.isDarwin [ "-DCMAKE_FRAMEWORK_PATH=${qwt}/lib" ];
```
I just got my NixOS VM back up so I'll look into it – it may be preferable to submit a patch upstream to fix the issue on both platforms.

And I might as well fix the style issues! I'll do it while I'm fixing the above.

Commit:
urh: platforms to linux only due to qt5 issue on Darwin

lukeadams commented:
@bjornfor I resolved the stylistic issues (whitespace and commit message related). I've also disabled that specific failing test on Linux. I'm rebuilding on Darwin and NixOS to be sure nothing broke after rebasing against master.

Sorry about the extensive reviewing work - I've learned so much from this!

bjornfor commented on pkgs/applications/misc/gnuradio/default.nix:
```diff
@@ -68,7 +68,7 @@ stdenv.mkDerivation rec {
         wrapProgram "$file" \
             --prefix PYTHONPATH : $PYTHONPATH:$(toPythonPath "$out") \
-            --set MATPLOTLIBRC "$out/share/gnuradio"
+            --set MATPLOTLIBRC "$out/share/gnuradio" ${stdenv.lib.optionalString stdenv.isDarwin " --set DYLD_FRAMEWORK_PATH /System/Library/Frameworks"}
```
Wrap this line?

```
--set MATPLOTLIBRC "$out/share/gnuradio" \
${stdenv.lib.optionalString stdenv.isDarwin ... }
```

bjornfor commented on pkgs/applications/misc/gnuradio-ais/default.nix:
```diff
@@ -12,8 +12,9 @@ stdenv.mkDerivation rec {
-    rev = "1863d1bf8a7709a8dfedb3ddb8e2b99112e7c872";
-    sha256 = "1vl3kk8xr2mh5lf31zdld7yzmwywqffffah8iblxdzblgsdwxfl6";
+    # Upstream PR: https://github.com/bistromath/gr-ais/commit/8502d0252a2a1a9b8d1a71795eaeb5d820684054
+    "rev" = "8502d0252a2a1a9b8d1a71795eaeb5d820684054";
```
Don't forget to update "version" above.

bjornfor replied:
It seems weird that version goes back one year. Can you clarify a bit in the commit message what's going on here? Also, perhaps make a point of switching from building upstream branch to a PR.

lukeadams replied:
Not sure where the 2016-08-26 date even originated from as the commit referenced (https://github.com/bistromath/gr-ais/commit/1863d1bf8a7709a8dfedb3ddb8e2b99112e7c872) is on the date I changed it to. 
But yeah I'll add both for clarification

bjornfor commented:
I started a build on NixOS (too), and ended up with this result:

```
...
99% tests passed, 1 tests failed out of 208

Total Test time (real) = 207.77 sec

The following tests FAILED:
        149 - qa_constellation_receiver (Failed)
Errors while running CTest
make: *** [Makefile:152: test] Error 8
builder for ‘/nix/store/6kkc2n98g942m1zapcl5693sdlgh5y95-gnuradio-3.7.11.drv’ failed with exit code 2
error: build of ‘/nix/store/6kkc2n98g942m1zapcl5693sdlgh5y95-gnuradio-3.7.11.drv’ failed
The invocation of "nix-build -A gnuradio-osmosdr -A gnuradio-with-packages -A inspectrum -A qwt6_qt4 -A gqrx -A gnuradio-rds -A gplates -A gnuradio-gsm -A gnuradio-nacl -A gnuradio -A libosmocore -A gnuradio-ais /home/bfo/proj/code/forks/nixpkgs-master" failed
```

lukeadams commented on pkgs/applications/misc/gnuradio-ais/default.nix:
```diff
@@ -12,8 +12,9 @@ stdenv.mkDerivation rec {
-    rev = "1863d1bf8a7709a8dfedb3ddb8e2b99112e7c872";
-    sha256 = "1vl3kk8xr2mh5lf31zdld7yzmwywqffffah8iblxdzblgsdwxfl6";
+    # Upstream PR: https://github.com/bistromath/gr-ais/commit/8502d0252a2a1a9b8d1a71795eaeb5d820684054
+    "rev" = "8502d0252a2a1a9b8d1a71795eaeb5d820684054";
```
[1863d1b](https://github.com/bistromath/gr-ais/commit/1863d1bf8a7709a8dfedb3ddb8e2b99112e7c872) (Dec 20, 2015) is the newest commit on the repo, which is "unversioned"
[8502d02](https://github.com/bistromath/gr-ais/pull/18/commits/8502d0252a2a1a9b8d1a71795eaeb5d820684054) is just my pr that fixes a linking issue on top of the above. 
Although the version didn't change, should I change the version to the correct date? ("2015-12-20", the last commit to master)

bjornfor replied:
Yes, version should be the commit date (for unversioned repositories).

lukeadams commented:
Hmm, that one (possibly along with a few others) is intermittent based on the amount of CPU usage since it basically generates large amounts of data and operates on the data stream. Although it didn't fail on my NixOS build, I'm going to go through and disable any tests that may cause non-deterministic behavior.

Commit:
gnuradio: force wxgui and pygtk check to true

enables gnuradio-companion compilation in headless/nix-multiuser environment

lukeadams commented:
@bjornfor 
Apparently gnuradio-companion wasn't being enabled in nix-multiuser environments due to the wx/pygtk imports failing. ead343b fixes this

I've added a "grSkipTests" list to allow simpler test skipping in the future, in addition to skipping tests which seem to fail regularly. (4494343)

Darwin succeeded without any issues after multiple runs. I'm testing on NixOS right now...

bjornfor commented:
@lukeadams: Ping me when the PR is ready.

bjornfor commented:
@lukeadams: "nox-review" passed for me now. Good to go?

lukeadams commented:
Almost! I rebased to the latest nixpkgs master and one of the python inputs broke. I need to hunt it down but I'll ping you when I do.

building with `nox-review ...` or `nix-shell -p gnuradio` (for e.g.) shows errors with python27Packages.repoze_lru and occasionally python27Packages.zope_lru
Not sure if it's just my configuration so the synced tree may work fine for you (hopefully it does!)

lukeadams commented:
@bjornfor I ended up rebasing on top of the commit before the gcc breakage so it's all good to go.

This package fails to build due to later commits to master (on Darwin), but its due a gcc breaking and not any of the packages in this PR. Once #27889 is fixed it will build fine.

lukeadams commented:
Current blockers for building on master:
#28385 (qt4)
[gfortran](https://hydra.nixos.org/build/57639613)

Commit:
gnuradio: 3.7.11.0 -> 3.7.11.1

this version includes GCC6 improvements

Commit:
gnuradio: Resolve darwin c++11 issues

-std=c++11 causes errors with OBJ-C files on Darwin. Inject dynamic compiler flag into every cmake file to work around this.

lukeadams commented:
@bjornfor @fpletz I just pushed my fixes for issues arising from cxx11 flags and GCC6. Merge conflicts have been resolved as well. Rebased onto master as of 11/11/17

bjornfor commented on pkgs/applications/misc/gnuradio/default.nix:
```diff
@@ -28,9 +28,11 @@ stdenv.mkDerivation rec {
   name = "gnuradio-${version}";
   version = "3.7.11";
```
Please update version attribute.

lukeadams replied:
👍 
I also changed it to use fetchFromGithub

lukeadams commented:
tested on NixOS + macOS 10.11

bjornfor commented:
Tested locally with nox-review and found some build regressions ("python3.6-hupper-1.0.drv’ failed with exit code 2", plus dependees). Can you have a look at that?

lukeadams commented:
seems to be a test suite inconsistency between the pip versions. Explicitly disabling tests fixed it but I still need to run now-review later today.

lukeadams commented:
@bjornfor I rebased this to include copumpkins hupper fix, removing the need for my commit. Also nice because it reduces the rebuild impact.

nox-review passes on darwin and NixOS.

bjornfor commented on pkgs/development/libraries/qwt/6_qt4.nix:
```diff
@@ -18,6 +18,13 @@ stdenv.mkDerivation rec {
   '';
 
+  # qwt.framework output includes a relative reference to itself, which breaks dependents
+  fixupPhase =
```
I think this should use preFixup (or postFixup), unless you really mean to override the fixupPhase. (In that case, a comment is required.)

lukeadams replied:
👍

bjornfor commented:
Nitpick: In commit messages with multiple lines, there should be a blank line separating the subject line from the body.

lukeadams commented:
@bjornfor I fixed the stylistic issues you mentioned. Let me know if there's anything else!

bjornfor commented:
@lukeadams: Thanks!

lukeadams commented:
@bjornfor Thanks for the guidance and help with this PR. I just got one of my workstations reconfigured and was happy to see that it's all in the binary cache for darwin, and it even works on High Sierra! ;)
